plugins/inputs/openldap/openldap.go

```diff
@@ -0,0 +1,148 @@
+
+import (
+	"crypto/tls"
+	"fmt"
+	"gopkg.in/ldap.v2"
+	"strconv"
+	"strings"
+
+	"github.com/influxdata/telegraf"
+	"github.com/influxdata/telegraf/plugins/inputs"
+)
+
+type Openldap struct {
+	Host          string
+	Port          int
+	Tls           bool
+	TlsSkipverify bool
+	BindDn        string
+	BindPassword  string
+}
+
+const sampleConfig string = `
+  host = "localhost"
+  port = 389
+  # starttls. Default is false.
+  tls = false
+  # skip peer certificate verification. Default is false.
+  tls_skipverify = false
+
+  # dn/password to bind with. If bind_dn is empty, an anonymous bind is performed.
+  bind_dn = ""
+  bind_password = ""
+`
+
+var searchBase = "cn=Monitor"
+var searchFilter = "(|(objectClass=monitorCounterObject)(objectClass=monitorOperation))"
+var searchAttrs = []string{"monitorCounter", "monitorOpInitiated", "monitorOpCompleted"}
+var attrTranslate = map[string]string{
+	"monitorCounter":     "",
+	"monitorOpInitiated": "_initiated",
+	"monitorOpCompleted": "_completed",
+}
+
+func (o *Openldap) SampleConfig() string {
+	return sampleConfig
+}
+
+func (o *Openldap) Description() string {
+	return "OpenLDAP cn=Monitor plugin"
+}
+
+// return an initialized Openldap
+func NewOpenldap() *Openldap {
+	return &Openldap{
+		Host:          "localhost",
+		Port:          389,
+		Tls:           false,
+		TlsSkipverify: false,
+		BindDn:        "",
+		BindPassword:  "",
+	}
+}
+
+// gather metrics
+func (o *Openldap) Gather(acc telegraf.Accumulator) error {
+	l, err := ldap.Dial("tcp", fmt.Sprintf("%s:%d", o.Host, o.Port))
+	if err != nil {
+		acc.AddError(err)
+		return nil
+	}
+	defer l.Close()
+
+	// TLS
+	if o.Tls {
+		err = l.StartTLS(&tls.Config{InsecureSkipVerify: o.TlsSkipverify})
+		if err != nil {
+			acc.AddError(err)
+			return nil
+		}
+	}
+
+	// username/password bind
+	if o.BindDn != "" && o.BindPassword != "" {
+		err = l.Bind(o.BindDn, o.BindPassword)
+		if err != nil {
+			acc.AddError(err)
+			return nil
+		}
+	}
+
+	searchRequest := ldap.NewSearchRequest(
+		searchBase,
+		ldap.ScopeWholeSubtree,
+		ldap.NeverDerefAliases,
+		0,
+		0,
+		false,
+		searchFilter,
+		searchAttrs,
+		nil,
+	)
+
+	sr, err := l.Search(searchRequest)
+	if err != nil {
+		acc.AddError(err)
+		return nil
+	}
+
+	gatherSearchResult(sr, o, acc)
+
+	return nil
+}
+
+func gatherSearchResult(sr *ldap.SearchResult, o *Openldap, acc telegraf.Accumulator) error {
+	fields := map[string]interface{}{}
+	tags := map[string]string{
+		"server": o.Host,
+		"port":   strconv.Itoa(o.Port),
+	}
+	for _, entry := range sr.Entries {
+		metricName := dnToMetric(entry.DN, searchBase)
+		for _, attr := range entry.Attributes {
+			if len(attr.Values[0]) >= 1 {
+				if v, err := strconv.ParseFloat(attr.Values[0], 64); err == nil {
+					fields[metricName+attrTranslate[attr.Name]] = v
+				}
+			}
+		}
+	}
+	acc.AddFields("openldap", fields, tags)
+	return nil
+}
+
+// Convert a DN to metric name, eg cn=Read,cn=Waiters,cn=Monitor to read_waiters
```

Comment:
I debated having a map of DNs to field names instead. It does bother my OCD just a bit having read_waiters instead of waiters_read, but it also reads a little easier? If other plugins name their fields so they sort more naturally I can change this one to match that pattern.